In [58]:
import pandas as pd
import numpy as np


In [100]:
import pandas as pd
import numpy as np
CSV_FILE_PATH='C:\\Users\\LiangShu\\Desktop\\final\\train_final.csv'
TEST_CSV_FILE_PATH='C:\\Users\\LiangShu\\Desktop\\final\\test_final.csv'
df=pd.read_csv(CSV_FILE_PATH)             #load train data
df_test=pd.read_csv(TEST_CSV_FILE_PATH)   #load test data
df.fillna(0,inplace=True)                         #fill NaN to 0
df_test.fillna(0,inplace=True)
df.head(3)

continuous_annual_inc  continuous_annual_inc_joint  continuous_delinq_2yrs  \
0                55000.0                          0.0                     0.0   
1                65000.0                          0.0                     1.0   
2                63000.0                      71000.0                     0.0   

   continuous_dti  continuous_dti_joint  continuous_fico_range_high  \
0            5.91                  0.00                       679.0   
1           16.06                  0.00                       719.0   
2           10.78                 13.85                       699.0   

   continuous_fico_range_low  continuous_funded_amnt  \
0                      675.0                  3600.0   
1                      715.0                 24700.0   
2                      695.0                 20000.0   

   continuous_funded_amnt_inv  continuous_inq_last_6mths  ...  \
0                      3600.0                        1.0  ...   
1                     24700.0                        4.0  ...   
2                     20000.0                        0.0  ...   

   discrete_sub_grade_28_one_hot  discrete_sub_grade_29_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_sub_grade_30_one_hot  discrete_sub_grade_31_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_sub_grade_32_one_hot  discrete_sub_grade_33_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_sub_grade_34_one_hot  discrete_sub_grade_35_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_term_1_one_hot  discrete_term_2_one_hot  
0                        1                        0  
1                        1                        0  
2                        0                        1  

[3 rows x 146 columns]

In [60]:
train_lable=df['loan_status']
test_lable=df_test['loan_status']
train_data=df.drop(['loan_status'],axis=1)
test_data=df_test.drop(['loan_status'],axis=1)
print(train_lable.sum())
train_data.head(3)

39788


continuous_annual_inc  continuous_annual_inc_joint  continuous_delinq_2yrs  \
0                55000.0                          0.0                     0.0   
1                65000.0                          0.0                     1.0   
2                63000.0                      71000.0                     0.0   

   continuous_dti  continuous_dti_joint  continuous_fico_range_high  \
0            5.91                  0.00                       679.0   
1           16.06                  0.00                       719.0   
2           10.78                 13.85                       699.0   

   continuous_fico_range_low  continuous_funded_amnt  \
0                      675.0                  3600.0   
1                      715.0                 24700.0   
2                      695.0                 20000.0   

   continuous_funded_amnt_inv  continuous_inq_last_6mths  ...  \
0                      3600.0                        1.0  ...   
1                     24700.0                        4.0  ...   
2                     20000.0                        0.0  ...   

   discrete_sub_grade_28_one_hot  discrete_sub_grade_29_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_sub_grade_30_one_hot  discrete_sub_grade_31_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_sub_grade_32_one_hot  discrete_sub_grade_33_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_sub_grade_34_one_hot  discrete_sub_grade_35_one_hot  \
0                              0                              0   
1                              0                              0   
2                              0                              0   

   discrete_term_1_one_hot  discrete_term_2_one_hot  
0                        1                        0  
1                        1                        0  
2                        0                        1  

[3 rows x 145 columns]

## 1- 传统模型
**模型1-SVM**

In [54]:
from sklearn import svm
clf=svm.SVC(kernel='rbf')
clf.fit(train_data,train_lable)
predict_trainData=clf.predict(train_data)


In [60]:
clf.score(train_data,train_lable)

0.79582

In [61]:
sum(predict_trainData)

49997

总共有50000个参数，39788个标签为1，标签为0的有10212个。但是预测结果为49997个为1，预测为0的仅仅只有3个。几乎没有分类

**模型2-Logistic regression**

In [61]:
from sklearn.linear_model import LogisticRegression
logis_clf=LogisticRegression(random_state=20)
logis_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(logis_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(logis_clf.score(test_data,test_lable)))

训练集评估结果：0.9107
测试集评估结果：0.9111


C:\Users\LiangShu\Anaconda3\envs\tfenv\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


**模型3-SGDClassifier**

In [62]:
from sklearn.linear_model import SGDClassifier
SGD_clf = SGDClassifier(random_state=29,loss='log')
SGD_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(SGD_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(SGD_clf.score(test_data,test_lable)))

训练集评估结果：0.88168
测试集评估结果：0.88548


**模型4-Gaussian Naive Bayes**

In [63]:
from sklearn.naive_bayes import GaussianNB
bayes_clf=GaussianNB()
bayes_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(bayes_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(bayes_clf.score(test_data,test_lable)))

训练集评估结果：0.90654
测试集评估结果：0.90478


**模型5-Decision Trees**

In [64]:
from sklearn import tree
tree_clf = tree.DecisionTreeClassifier()
tree_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(tree_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(tree_clf.score(test_data,test_lable)))

训练集评估结果：1.0
测试集评估结果：0.88058


**模型6-Ensemble methods**
1. Bagging meta-estimator
2. Forests of randomized trees
3. ExtraTreesClassifier
4. AdaBoost
5. Gradient Tree Boosting
6. Histogram-Based Gradient Boosting
7. Voting Classifier
8. Voting Regressor
9. Stacked generalization

In [65]:
# Bagging meta-estimator
from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
bagging = BaggingClassifier(KNeighborsClassifier(),max_samples=0.6, max_features=0.6)
bagging.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(bagging.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(bagging.score(test_data,test_lable)))

训练集评估结果：0.89548
测试集评估结果：0.8667


In [66]:
# Forests of randomized trees
from sklearn.ensemble import RandomForestClassifier
randomforest_clf=RandomForestClassifier(n_estimators=10)
randomforest_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(randomforest_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(randomforest_clf.score(test_data,test_lable)))

训练集评估结果：0.99434
测试集评估结果：0.91154


In [67]:
# ExtraTreesClassifier
from sklearn.ensemble import ExtraTreesClassifier
extratree_clf=ExtraTreesClassifier(n_estimators=10)
extratree_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(extratree_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(extratree_clf.score(test_data,test_lable)))

训练集评估结果：1.0
测试集评估结果：0.90694


In [68]:
# AdaBoost
from sklearn.ensemble import AdaBoostClassifier
adam_clf = AdaBoostClassifier(n_estimators=100,learning_rate=0.88)
adam_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(adam_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(adam_clf.score(test_data,test_lable)))

训练集评估结果：0.91996
测试集评估结果：0.91592


In [69]:
# Gradient Tree Boosting
from sklearn.ensemble import GradientBoostingClassifier
gbc_clf = GradientBoostingClassifier(n_estimators=100, learning_rate=0.88, random_state=0)
gbc_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(gbc_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(gbc_clf.score(test_data,test_lable)))

训练集评估结果：0.93192
测试集评估结果：0.91346


In [71]:
# Histogram-Based Gradient Boosting, can't loaded this method, ignore it
# from sklearn.ensemble import HistGradientBoostingClassifier
# Hist_clf = HistGradientBoostingClassifier(max_iter=100)
# Hist_clf.fit(train_data,train_lable)
# #评估结果
# print('训练集评估结果：{}'.format(Hist_clf.score(train_data,train_lable)))
# print('测试集评估结果：{}'.format(Hist_clf.score(test_data,test_lable)))

ImportError: cannot import name 'HistGradientBoostingClassifier' from 'sklearn.ensemble' (C:\Users\LiangShu\Anaconda3\envs\tfenv\lib\site-packages\sklearn\ensemble\__init__.py)

**模型6-lightGBM**

In [72]:
from lightgbm import LGBMClassifier
lgb_clf=LGBMClassifier()
lgb_clf.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(lgb_clf.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(lgb_clf.score(test_data,test_lable)))

训练集评估结果：0.93234
测试集评估结果：0.91754


# 2- stacking合成
由上面的不同模型可以看出，lightGBM, Gradient Tree Boosting, AdaBoost, Forests of randomized trees, ExtraTreesClassifier这五个模型效果较好，所以选择他们进行集成。
1. 将数据分成5块然后用于不同模型训练
2. 五个模型分别计算

In [77]:
from lightgbm import LGBMClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import StackingClassifier

randomforest_clf=RandomForestClassifier(n_estimators=10)
adam_clf = AdaBoostClassifier(n_estimators=100,learning_rate=0.88)
gbc_clf = GradientBoostingClassifier(n_estimators=100, learning_rate=0.88, random_state=0)
lgb_clf=LGBMClassifier()
extratree_clf=ExtraTreesClassifier(n_estimators=10)


estimators = [('ExtraTreesClassifier', extratree_clf),
              ('randomforest', randomforest_clf),
              ('AdamBoost', adam_clf),
              ('GradientBoost', gbc_clf),
              ('LightGBM', lgb_clf)]

# Choose lightGBM as final estimator as the score is the best
final_estimator =LGBMClassifier()

# stacking Regressors
reg = StackingClassifier(
        estimators=estimators,
        final_estimator=final_estimator)

reg.fit(train_data,train_lable)
#评估结果
print('训练集评估结果：{}'.format(reg.score(train_data,train_lable)))
print('测试集评估结果：{}'.format(reg.score(test_data,test_lable)))



训练集评估结果：0.92636
测试集评估结果：0.9165


### 3- 网络模型，tabnet

In [78]:
!pip install pytorch-tabnet

You should consider upgrading via the 'c:\users\liangshu\anaconda3\envs\tfenv\python.exe -m pip install --upgrade pip' command.


In [117]:
from pytorch_tabnet.tab_model import TabNetClassifier, TabNetRegressor
X_train=train_data.loc[0:40000].values
Y_train=train_lable.loc[0:40000].values
X_valid=train_data.loc[40000:50000].values
Y_valid=train_lable.loc[40000:50000].values


In [118]:
tabnet_clf=TabNetClassifier()
tabnet_clf.fit(X_train,Y_train,
               max_epochs=50,
               eval_set=[(X_train, Y_train),(X_valid, Y_valid)],
               eval_name=['train', 'valid'],
               eval_metric=['auc'])

Device used : cpu
epoch 0  | loss: 0.43028 | train_auc: 0.50276 | valid_auc: 0.50668 |  0:00:09s
epoch 1  | loss: 0.24737 | train_auc: 0.86322 | valid_auc: 0.86457 |  0:00:20s
epoch 2  | loss: 0.22659 | train_auc: 0.92071 | valid_auc: 0.91884 |  0:00:32s
epoch 3  | loss: 0.22123 | train_auc: 0.93574 | valid_auc: 0.93351 |  0:00:43s
epoch 4  | loss: 0.21935 | train_auc: 0.95348 | valid_auc: 0.9501  |  0:00:54s
epoch 5  | loss: 0.21686 | train_auc: 0.95328 | valid_auc: 0.94922 |  0:01:05s
epoch 6  | loss: 0.21229 | train_auc: 0.95473 | valid_auc: 0.95094 |  0:01:16s
epoch 7  | loss: 0.20984 | train_auc: 0.9568  | valid_auc: 0.95385 |  0:01:28s
epoch 8  | loss: 0.20925 | train_auc: 0.95803 | valid_auc: 0.9552  |  0:01:40s
epoch 9  | loss: 0.21032 | train_auc: 0.95758 | valid_auc: 0.95353 |  0:01:53s
epoch 10 | loss: 0.2091  | train_auc: 0.95824 | valid_auc: 0.95502 |  0:02:04s
epoch 11 | loss: 0.20726 | train_auc: 0.95872 | valid_auc: 0.95602 |  0:02:15s
epoch 12 | loss: 0.20692 | train_a

In [120]:
from sklearn.metrics import accuracy_score
X_test=test_data.values
Y_test=test_lable.values
preds = tabnet_clf.predict(X_test)

accuracy_score(Y_test,preds)

0.91366

### 4- Trik
**修改mask函数**

In [124]:
tabnet_clf2=TabNetClassifier(mask_type='entmax')
tabnet_clf2.fit(X_train,Y_train,
               max_epochs=50,
               eval_set=[(X_train, Y_train),(X_valid, Y_valid)],
               eval_name=['train', 'valid'],
               eval_metric=['auc'])

Device used : cpu
epoch 0  | loss: 0.38423 | train_auc: 0.55959 | valid_auc: 0.55478 |  0:00:10s
epoch 1  | loss: 0.22713 | train_auc: 0.82522 | valid_auc: 0.82666 |  0:00:20s
epoch 2  | loss: 0.21191 | train_auc: 0.85566 | valid_auc: 0.85042 |  0:00:31s
epoch 3  | loss: 0.20842 | train_auc: 0.91118 | valid_auc: 0.9065  |  0:00:43s
epoch 4  | loss: 0.20981 | train_auc: 0.91152 | valid_auc: 0.90636 |  0:00:53s
epoch 5  | loss: 0.20617 | train_auc: 0.95195 | valid_auc: 0.94722 |  0:01:04s
epoch 6  | loss: 0.20668 | train_auc: 0.95516 | valid_auc: 0.95179 |  0:01:15s
epoch 7  | loss: 0.20533 | train_auc: 0.95925 | valid_auc: 0.95642 |  0:01:26s
epoch 8  | loss: 0.20408 | train_auc: 0.96043 | valid_auc: 0.95735 |  0:01:37s
epoch 9  | loss: 0.20412 | train_auc: 0.96103 | valid_auc: 0.95827 |  0:01:48s
epoch 10 | loss: 0.20289 | train_auc: 0.96196 | valid_auc: 0.95786 |  0:01:59s
epoch 11 | loss: 0.20232 | train_auc: 0.9617  | valid_auc: 0.9571  |  0:02:11s
epoch 12 | loss: 0.20066 | train_a

In [125]:
preds2 = tabnet_clf2.predict(X_test)

accuracy_score(Y_test,preds2)

0.9142

**修改momentum参数**

In [132]:
import torch
tabnet_clf3=TabNetClassifier(mask_type='entmax',momentum=0.01)
tabnet_clf3.fit(X_train,Y_train,
               max_epochs=30,
               eval_set=[(X_train, Y_train),(X_valid, Y_valid)],
               eval_name=['train', 'valid'],
               eval_metric=['auc'])

Device used : cpu
epoch 0  | loss: 0.38423 | train_auc: 0.56178 | valid_auc: 0.5575  |  0:00:09s
epoch 1  | loss: 0.22713 | train_auc: 0.79845 | valid_auc: 0.80433 |  0:00:20s
epoch 2  | loss: 0.21191 | train_auc: 0.84416 | valid_auc: 0.84176 |  0:00:30s
epoch 3  | loss: 0.20842 | train_auc: 0.90564 | valid_auc: 0.90066 |  0:00:42s
epoch 4  | loss: 0.20981 | train_auc: 0.90656 | valid_auc: 0.90064 |  0:00:53s
epoch 5  | loss: 0.20617 | train_auc: 0.95148 | valid_auc: 0.94667 |  0:01:06s
epoch 6  | loss: 0.20668 | train_auc: 0.95475 | valid_auc: 0.95154 |  0:01:19s
epoch 7  | loss: 0.20533 | train_auc: 0.95922 | valid_auc: 0.95643 |  0:01:31s
epoch 8  | loss: 0.20408 | train_auc: 0.96046 | valid_auc: 0.95733 |  0:01:44s
epoch 9  | loss: 0.20412 | train_auc: 0.961   | valid_auc: 0.95823 |  0:01:55s
epoch 10 | loss: 0.20289 | train_auc: 0.96195 | valid_auc: 0.95783 |  0:02:06s
epoch 11 | loss: 0.20232 | train_auc: 0.96171 | valid_auc: 0.95701 |  0:02:17s
epoch 12 | loss: 0.20066 | train_a

In [134]:
preds3 = tabnet_clf3.predict(X_test)

accuracy_score(Y_test,preds3)

0.91438

**修改optimizer_params**

In [145]:
import torch
import torch.nn

tabnet_clf4=TabNetClassifier(mask_type='entmax',momentum=0.01,optimizer_params=dict(lr=0.01))
tabnet_clf4.fit(X_train,Y_train,
               max_epochs=30,
               eval_set=[(X_train, Y_train),(X_valid, Y_valid)],
               eval_name=['train', 'valid'],
               eval_metric=['auc'])

Device used : cpu
epoch 0  | loss: 0.46429 | train_auc: 0.73367 | valid_auc: 0.7523  |  0:00:11s
epoch 1  | loss: 0.2594  | train_auc: 0.89742 | valid_auc: 0.89016 |  0:00:22s
epoch 2  | loss: 0.23041 | train_auc: 0.91893 | valid_auc: 0.91907 |  0:00:33s
epoch 3  | loss: 0.22198 | train_auc: 0.92522 | valid_auc: 0.92539 |  0:00:45s
epoch 4  | loss: 0.21623 | train_auc: 0.95183 | valid_auc: 0.95135 |  0:00:58s
epoch 5  | loss: 0.21293 | train_auc: 0.95565 | valid_auc: 0.95396 |  0:01:11s
epoch 6  | loss: 0.21139 | train_auc: 0.95787 | valid_auc: 0.95568 |  0:01:24s
epoch 7  | loss: 0.20873 | train_auc: 0.95867 | valid_auc: 0.95656 |  0:01:38s
epoch 8  | loss: 0.20756 | train_auc: 0.95942 | valid_auc: 0.95688 |  0:01:54s
epoch 9  | loss: 0.20767 | train_auc: 0.96043 | valid_auc: 0.95693 |  0:02:09s
epoch 10 | loss: 0.2073  | train_auc: 0.96078 | valid_auc: 0.95716 |  0:02:22s
epoch 11 | loss: 0.206   | train_auc: 0.96079 | valid_auc: 0.9571  |  0:02:34s
epoch 12 | loss: 0.20508 | train_a

In [146]:
preds4 = tabnet_clf4.predict(X_test)

accuracy_score(Y_test,preds4)

0.9167

最终score为0.9167